meetings/working-groups/field-keyword/FK-2024-08-07.md

```diff
@@ -0,0 +1,125 @@
+
+```cs
+class C
+{
+    public C() { Prop = "..." };
+
+    public string Prop
+    {
+        get;
+        set
+        {
+            if (condition) field = value;
+        //  CS8774 Member 'field' must have a non-null value when exiting.
+        }
+        ~
+    }
+}
+```
+
+### Nullable backing field
+
+We preferred the idea of making the backing field nullable rather than keeping it matching the property type while applying `[field: AllowNull, MaybeNull]`. Users are by far more familiar with using nullable types than working with the attributes. The null-resilient getter concept is stronger when its usage and explanation is simpler for users. Making the field nullable at the symbolic level is a difference that would be visible through the IDE, where the user would see a question mark after the type when hovering over the `field` keyword.
+
+### `[AllowNull]` nullability hole
+
+The proposal mentions that there is an existing nullability hole in the language when `[AllowNull]` is applied to a field, and that this hole will be inherited by `[field: AllowNull]`:
+
+```cs
+// C# 12
+class C
+{
+    [AllowNull]
+    private string resetIfSetToDefault;
+
+    public string ResetIfSetToDefault
+    {
+        // Removed `?? GetDefault()` without warnings!
+        get => resetIfSetToDefault; 
+        set => resetIfSetToDefault = value == GetDefault() ? null : value;
+    }
+}
+```
+
+The point was raised that when it comes to `[field: AllowNull]`, we might have an opportunity to close that gap in warnings because the backing field is scoped to the property. For instance, perhaps we could simply create a nullability warning if `[field: AllowNull]` is used without `[field: MaybeNull]`.
+
+As mentioned elsewhere, `[field: AllowNull]` is likely only going to be specified in advanced scenarios where automatic backing field nullability (somewhat equivalent to AllowNull and MaybeNull together) isn't what the user wants.
+
+### Open question: unrelated nullability warnings
+
+We considered whether unrelated nullability warnings should affect the nullability of the backing field:
+
+```cs
+// Is this a null-resilient getter?
+get
+{
+    string unrelated = null; // Warning
+    return field ?? "";
+}
+```
+
+It would likely not be expensive to detect relatedness. It would consist of doing two nullability analysis passes, only considering warnings that were present in a maybe-null-field pass and not present in a non-null-field pass. There could be occasional scenarios with inferred generic type arguments where there is a different error in each pass; however, what would fall out in that case would be a "not null-resilient" detection, which is a safe place to land, still closer to the mark than doing only a single pass. Finally, it might be harder to explain and understand "are there new warnings caused by `field` becoming nullable" compared to understanding "are there any warnings when `field` is nullable."
```

Comment:
> It would likely not be expensive to detect relatedness.

This is not the impression I got from the meeting; it seemed like most of us were in agreement that we weren't sure how complex it would be.

Reply:
Good clarification, I'll reword. If I remember right, it was the operation of diffing diagnostics between two passes that was considered to probably not be expensive. Then we weren't sure whether we would end up wanted to do something even more nuanced than that. (I think probably not.)

Reply:
We could diff fairly easily, but we weren't sure that would actually produce correct results.